File: Source/AlphaTab/Rendering/Glyphs/ScoreTieGlyph.cs
using AlphaTab.Model;
using AlphaTab.Rendering.Utils;

namespace AlphaTab.Rendering.Glyphs
{
    internal class ScoreTieGlyph : TieGlyph
    {
        protected readonly Note StartNote;
        protected readonly Note EndNote;

        public ScoreTieGlyph(Note startNote, Note endNote, bool forEnd = false)
            : base(startNote == null ? null : startNote.Beat, endNote == null ? null : endNote.Beat, forEnd)
        {
            StartNote = startNote;
            EndNote = endNote;
        }

        public override void DoLayout()
        {
            base.DoLayout();
            YOffset = NoteHeadGlyph.NoteHeadHeight / 2;
        }

        protected override BeamDirection GetBeamDirection(Beat beat, BarRendererBase noteRenderer)
        {
            // invert direction (if stems go up, ties go down to not cross them)
            switch (((ScoreBarRenderer)noteRenderer).GetBeatDirection(beat))
            {
                case BeamDirection.Up:
                    return BeamDirection.Down;
                case BeamDirection.Down:
                default:
                    return BeamDirection.Up;
            }
        }

        protected override float GetStartY(BarRendererBase noteRenderer, BeamDirection direction)
        {
            return noteRenderer.GetNoteY(StartNote);
        }

        protected override float GetEndY(BarRendererBase noteRenderer, BeamDirection direction)
        {
            return noteRenderer.GetNoteY(EndNote);
        }

        protected override float GetStartX(BarRendererBase noteRenderer)
        {
            return noteRenderer.GetBeatX(StartNote.Beat, BeatXPosition.MiddleNotes);
        }

        protected override float GetEndX(BarRendererBase noteRenderer)
        {
            return noteRenderer.GetNoteX(EndNote, false);
        }
    }
}

File: src/rendering/glyphs/ScoreTieGlyph.ts
import type { Beat } from '@src/model/Beat';
import type { Note } from '@src/model/Note';
import { type BarRendererBase, NoteYPosition } from '@src/rendering/BarRendererBase';
import { TieGlyph } from '@src/rendering/glyphs/TieGlyph';
import type { ScoreBarRenderer } from '@src/rendering/ScoreBarRenderer';
import { BeamDirection } from '@src/rendering/utils/BeamDirection';
import { BeatXPosition } from '@src/rendering/BeatXPosition';

export class ScoreTieGlyph extends TieGlyph {
    protected startNote: Note;
    protected endNote: Note;

    public constructor(startNote: Note, endNote: Note, forEnd: boolean = false) {
        super(!startNote ? null : startNote.beat, !endNote ? null : endNote.beat, forEnd);
        this.startNote = startNote;
        this.endNote = endNote;
    }

    protected override shouldDrawBendSlur() {
        return (
            this.renderer.settings.notation.extendBendArrowsOnTiedNotes &&
            !!this.startNote.bendOrigin &&
            this.startNote.isTieOrigin
        );
    }

    public override doLayout(): void {
        super.doLayout();
    }

    protected override getBeamDirection(beat: Beat, noteRenderer: BarRendererBase): BeamDirection {
        // invert direction (if stems go up, ties go down to not cross them)
        switch ((noteRenderer as ScoreBarRenderer).getBeatDirection(beat)) {
            case BeamDirection.Up:
                return BeamDirection.Down;
            default:
                return BeamDirection.Up;
        }
    }

    protected override getStartY(): number {
        if (this.startBeat!.isRest) {
            // below all lines
            return (this.startNoteRenderer as ScoreBarRenderer).getScoreY(9);
        }
        switch (this.tieDirection) {
            case BeamDirection.Up:
                // below lowest note
                return this.startNoteRenderer!.getNoteY(this.startNote, NoteYPosition.Top);
            default:
                return this.startNoteRenderer!.getNoteY(this.startNote, NoteYPosition.Bottom);
        }
    }

    protected override getEndY(): number {
        const endNoteScoreRenderer = this.endNoteRenderer as ScoreBarRenderer;
        if (this.endBeat!.isRest) {
            switch (this.tieDirection) {
                case BeamDirection.Up:
                    return endNoteScoreRenderer.getScoreY(9);
                default:
                    return endNoteScoreRenderer.getScoreY(0);
            }
        }

        switch (this.tieDirection) {
            case BeamDirection.Up:
                return endNoteScoreRenderer.getNoteY(this.endNote, NoteYPosition.Top);
            default:
                return endNoteScoreRenderer.getNoteY(this.endNote, NoteYPosition.Bottom);
        }
    }

    protected override getStartX(): number {
        return this.startNoteRenderer!.getBeatX(this.startNote.beat, BeatXPosition.PostNotes);
    }

    protected override getEndX(): number {
        return this.endNoteRenderer!.getBeatX(this.endNote.beat, BeatXPosition.PreNotes);
    }
}
